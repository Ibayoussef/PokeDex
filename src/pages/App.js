import React, { Component } from "react";
import "../css/App.css";
import Pokemon from "../components/Pokemon";
import PokemonModal from "../components/PokemonModal";
import Nav from "../components/Nav";
import InfiniteScroll from "react-infinite-scroll-component";
import axios from "axios";
export default class App extends Component {
  state = {
    pokemons: [],
    pokemonDetails: [],
    isFetching: false,
    offset: 0,
    loadNumber: 15,
    showModal: false,
    singlePokemon: [],
  };
  showModal = async (index) => {
    await axios
      .get(`https://pokeapi.co/api/v2/pokemon/${index}`)
      .then((res) => {
        const pokemon = res.data;
        this.setState({ singlePokemon: pokemon });
      });
    this.setState({ show: true });
  };

  hideModal = () => {
    this.setState({ show: false });
  };
  getNextOffset = () => {
    return this.state.offset + this.state.loadNumber;
  };

  handleMore = () => {
    const newOffset = this.getNextOffset();
    this.setState({ offset: newOffset }, () => {
      console.log("Offset: " + this.state.offset);
      this.getMorePokemon();
    });
  };
  componentDidMount() {
    this.getMorePokemon();
  }
  async getMorePokemon() {
    let url =
      "https://pokeapi.co/api/v2/pokemon?offset=" +
      this.state.offset +
      "&limit=" +
      this.state.loadNumber;
    await fetch(url)
      .then((response) => response.json())
      .then((data) => {
        if (data) {
          this.setState({ isFetching: true });
          this.setState({ pokemons: data.results }, () => {
            this.state.pokemons.map((pokemon) => {
              fetch(pokemon.url)
                .then((response) => response.json())
                .then((data) => {
                  if (data) {
                    var temp = this.state.pokemonDetails;
                    temp.push(data);
                    this.setState({ pokemonDetails: temp, isFetching: false });
                  }
                })
                .catch(console.log);
            });
          });
        }
      })
      .catch(console.log);
  }

  render() {
    const { pokemonDetails } = this.state;

    const renderedPokemonList = pokemonDetails.map((pokemon, index) => {
      return (
        <div>
          <PokemonModal
            show={this.state.show}
            pokemon={this.state.singlePokemon}
            handleClose={this.hideModal}
            key={this.state.singlePokemon.id}
          ></PokemonModal>
          <div onClick={this.showModal.bind(this, pokemon.id)}>
            <Pokemon pokemon={pokemon} key={pokemon.id} />
          </div>
        </div>
      );
    });

    return (
      <>
        <Nav />
        <div data-testid="App" className="container">
          <InfiniteScroll
            dataLength={this.state.pokemonDetails.length}
            next={this.handleMore}
            hasMore={true}
            loader={<h4>Loading...</h4>}
          >
            {this.state.isFetching ? (
              <div className="App">
                <h1>LOADING....</h1>
              </div>
            ) : (
              <div className="App">{renderedPokemonList}</div>
            )}
          </InfiniteScroll>
        </div>
      </>
    );
  }
}
