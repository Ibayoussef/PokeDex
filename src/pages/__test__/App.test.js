import React from "react";
import ReactDOM from "react-dom";
import App from "../App";
import { isTSAnyKeyword } from "@babel/types";
import { render, cleanup } from "@testing-library/react";
import renderer from "react-test-renderer";
import "@testing-library/jest-dom";

afterEach(cleanup);

it("renders without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<App></App>, div);
});

it("renders correctly", () => {
  const { getByTestId } = render(<App></App>);
  expect(getByTestId("App")).toHaveTextContent("Loading...");
});

it("matches snapshot", () => {
  const tree = renderer.create(<App></App>).toJSON();
  expect(tree).toMatchSnapshot();
});
