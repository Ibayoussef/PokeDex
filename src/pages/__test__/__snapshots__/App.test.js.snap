// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
Array [
  <div
    className="nav"
  >
    <div
      className="nav-header"
    >
      <div
        className="nav-title"
      >
        Pok√©Dex
      </div>
    </div>
  </div>,
  <div
    className="title"
  >
    <h1>
      Select A Pokemon
    </h1>
  </div>,
  <div
    className="container"
    data-testid="App"
  >
    <div
      className="infinite-scroll-component__outerdiv"
      style={Object {}}
    >
      <div
        className="infinite-scroll-component "
        style={
          Object {
            "WebkitOverflowScrolling": "touch",
            "height": "auto",
            "overflow": "auto",
          }
        }
      >
        <div
          className="App"
        />
        <h4>
          Loading...
        </h4>
      </div>
    </div>
  </div>,
]
`;
